generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -172,6 +173,74 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by taking the dot product of two vectors. This
+   * function depends on two rules of the dot product. Let u and v be vectors of the same size,
+   * and let i be a valid index into those vectors.
+   *     If u[i] != 0 <==> v[i] = 0, then dot(u, v) = 0. Example: u = (0, 1), v = (1, 0), then
+   *     dot(u, v) = 1 * 0 + 0 * 1 = 0.
+   *     If u[i] = v[i] = 0 for all i except for a single index, j, and if u[j] = v[j] = 1, then
+   *     dot(u, v) = 0. Example: u = (0, 1), v = (0, 1), then dot(u, v) = 0 * 0 + 1 * 1 = 1.
+   * Because fuzzed expressions are not always well defined, we forgo those in favor of simply
+   * using opaque ones and zeroes when generating vectors for this function.
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from the dot product of two vectors.
+   */
+  private Optional<Expr> opaqueZeroOrOneDot(BasicType type, boolean constContext,
+                                            final int depth, Fuzzer fuzzer, boolean isZero) {
+    if (type != BasicType.FLOAT) {
+      return Optional.empty();
+    }
+    // Vector sizes need to be in [2, 4].
+    final int vectorSize = generator.nextInt(3) + 2;
```

Comment:
It'd be better if I didn't use magic numbers here, but I couldn't think of what to name the constants, since they're a little arbitrary. Let me know if I should pursue that.

Reply:
The "dot" function is defined on float, as well as vec2--vec4, so we may as well allow this transformation to introduce "dot" expressions on float types.  So I suggest:

final int vectorSize = generator.nextPositiveInt(BasicType.VEC4.getNumElements() + 1);

Reply:
I didn't think of using BasicType.VEC4.getNumElements(), that's super self documenting - thanks.